lightning/src/ln/channelmanager.rs

```diff
@@ -3055,6 +3088,63 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 		Ok(())
 	}
 
+	/// Fails the intercepted payment indicated by intercept_id.  This should really only be called in response
+	/// to a PaymentIntercepted event
+	pub fn fail_intercepted_payment(&self, intercept_id: InterceptId) {
+		let _persistence_guard = PersistenceNotifierGuard::notify_on_drop(&self.total_consistency_lock, &self.persistence_notifier);
+
+		let pending_intercept = {
+			let mut pending_intercepts = self.pending_intercepted_payments.lock().unwrap();
+			pending_intercepts.remove(&intercept_id)
+		};
+
+		if let Some(payment) = pending_intercept {
+			let htlc_source = HTLCSource::PreviousHopData(HTLCPreviousHopData {
+				short_channel_id: payment.prev_short_channel_id,
+				outpoint: payment.prev_funding_outpoint,
+				htlc_id: payment.prev_htlc_id,
+				incoming_packet_shared_secret: payment.forward_info.incoming_shared_secret,
+				phantom_shared_secret: None,
+			});
+
+			let failure_reason = HTLCFailReason::Reason { failure_code: 0x4000 | 10, data: Vec::new() };
+			self.fail_htlc_backwards_internal(self.channel_state.lock().unwrap(), htlc_source, &payment.forward_info.payment_hash, failure_reason);
+		}
+	}
+
+	/// Attempts to forward an intercepted payment over the provided scid and with the provided amt_to_forward.
+	/// Should only really be called in response to a PaymentIntercepted event
+	pub fn forward_intercepted_payment(&self, intercept_id: InterceptId, scid: u64, amt_to_forward: u64) -> Result<(), APIError> {
+		let _persistence_guard = PersistenceNotifierGuard::notify_on_drop(&self.total_consistency_lock, &self.persistence_notifier);
+
+		let pending_intercept = {
+			let mut pending_intercepts = self.pending_intercepted_payments.lock().unwrap();
+			pending_intercepts.remove(&intercept_id)
+		};
+
+		match pending_intercept {
+			None => Err(APIError::APIMisuseError { err: "Payment with that InterceptId not found".to_string() }),
+			Some(payment) => {
+					let routing = match payment.forward_info.routing {
+						PendingHTLCRouting::Forward { onion_packet, .. } => {
+							PendingHTLCRouting::Forward { onion_packet, short_channel_id: scid }
+						},
+						_ => payment.forward_info.routing
```

Comment:
yeah, my gut is that we might as well support it. I assume this means surfacing the original onion in the event in addition to being able to pass a new one here, right?

Reply:
oh right, that makes sense. 

weren't there other non-lsp use-cases mentioned in here ( https://github.com/lightningdevkit/rust-lightning/issues/680 ) for a feature like this?

also it seems like this can still be useful for non-mpp supporting lsp.  i guess it really comes down to what trade-offs are being made with the mpp supporting lsp and if there's a strong case to support both flows.

Reply:
oh i guess i misunderstood 680 as well. i thought it alluded to being able surface htlcs to end-user similarly to this 'intercept' pr to enable them to add additional relay logic at the user-level.